### PR TITLE
Process blobs in VecSimIndexInterface functions - MOD-5163

### PR DESCRIPTION
**Pull request by meiravgri**
**Describe the changes in the pull request**
Instead of internally processing vectors/queries (currently, the only processing step we perform is normalizing the blobs), we added a process blob step to all the functions of `VecSimIndexInterface` that get a blob and send it to the internal index implementation.
In this PR we `VecSimIndexInterface` implements *Wrapper functions for the following functions:
 * `addVector`
 * `topKQuery`
 * `rangeQuery`
 * `newBatchIterator`
Those assume that they get the processed vector, and should be implemented internally by the indices that inherit from the `VecSimIndexInterface` class.

**Which issues this PR fixes**
1. When using the tiered index, we applied the normalization function twice - once in the flat buffer and then in the hnsw index.

**Main objects this PR modified**
1. split `VecSimIndexInterface` functions into a wrapper function and implementation function.

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: src/VecSim/vec_sim_interface.h [hunk 1]
```diff
@@ -33,11 +33,21 @@ struct VecSimIndexInterface : public VecsimBaseObject {
...
+     * It is assumed the the index saves its own copy of the blob.
+     */
+    int addVectorWrapper(const void *blob, labelType label, void *auxiliaryCtx = nullptr);
+
     /**
      * @brief Add a vector blob and its id to the index.
      *
      * @param blob binary representation of the vector. Blob size should match the index data type
-     * and dimension.
+     * and dimension.  It is assumed that the queryBlob has been already processed
```
Comment (GuyAv46) [inline src/VecSim/vec_sim_interface.h:49]:
```suggestion
     * and dimension. It is assumed that the queryBlob has been already processed
```

**Review by GuyAv46 — APPROVED** [review]

File: src/VecSim/vec_sim_index.h [hunk 2]
```diff
@@ -98,4 +103,48 @@ struct VecSimIndexAbstract : public VecSimIndexInterface {
...
+        return original_blob;
+    }
+
+protected:
+    virtual int addVectorWrapper(const void *blob, labelType label, void *auxiliaryCtx) override {
+        char processed_blob[this->data_size];
+        const void *vector_to_add = processBlob(blob, processed_blob);
+
+        return this->addVector(vector_to_add, label, auxiliaryCtx);
+    }
```
Comment (DvirDukhan) [inline src/VecSim/vec_sim_index.h:128]: new line

File: src/VecSim/vec_sim_index.h [hunk 3]
```diff
@@ -98,4 +103,48 @@ struct VecSimIndexAbstract : public VecSimIndexInterface {
...
+    virtual VecSimQueryResult_List topKQueryWrapper(const void *queryBlob, size_t k,
+                                                    VecSimQueryParams *queryParams) override {
+        char processed_blob[this->data_size];
+        const void *query_to_send = processBlob(queryBlob, processed_blob);
+
+        return this->topKQuery(query_to_send, k, queryParams);
+    }
+    virtual VecSimQueryResult_List rangeQueryWrapper(const void *queryBlob, double radius,
```
Comment (DvirDukhan) [inline src/VecSim/vec_sim_index.h:136]: new line

File: src/VecSim/vec_sim_index.h [hunk 4]
```diff
@@ -98,4 +103,48 @@ struct VecSimIndexAbstract : public VecSimIndexInterface {
...
+                                                     VecSimQueryParams *queryParams) override {
+        char processed_blob[this->data_size];
+        const void *query_to_send = processBlob(queryBlob, processed_blob);
+
+        return this->rangeQuery(query_to_send, radius, queryParams);
+    }
+    virtual VecSimBatchIterator *
```
Comment (DvirDukhan) [inline src/VecSim/vec_sim_index.h:143]: new line

File: src/VecSim/vec_sim_tiered_index.h [hunk 5]
```diff
@@ -59,6 +59,37 @@ class VecSimTieredIndex : public VecSimIndexInterface {
...
                                      VecSimQueryParams *queryParams) override;
+
+private:
+    virtual int addVectorWrapper(const void *blob, labelType label, void *auxiliaryCtx) override {
+        // Will be used only if a processing stage is needed
+        char processed_blob[this->backendIndex->getDataSize()];
+        const void *vector_to_add = this->backendIndex->processBlob(blob, processed_blob);
+        return this->addVector(vector_to_add, label, auxiliaryCtx);
+    }
+    virtual VecSimQueryResult_List topKQueryWrapper(const void *queryBlob, size_t k,
```
Comment (DvirDukhan) [inline src/VecSim/vec_sim_tiered_index.h:70]: new line

File: src/VecSim/vec_sim_tiered_index.h [hunk 6]
```diff
@@ -59,6 +59,37 @@ class VecSimTieredIndex : public VecSimIndexInterface {
...
+                                                    VecSimQueryParams *queryParams) override {
+        // Will be used only if a processing stage is needed
+        char processed_blob[this->backendIndex->getDataSize()];
+        const void *query_to_send = this->backendIndex->processBlob(queryBlob, processed_blob);
+        return this->topKQuery(query_to_send, k, queryParams);
+    }
+    virtual VecSimQueryResult_List rangeQueryWrapper(const void *queryBlob, double radius,
```
Comment (DvirDukhan) [inline src/VecSim/vec_sim_tiered_index.h:77]: new line

File: src/VecSim/vec_sim_tiered_index.h [hunk 7]
```diff
@@ -59,6 +59,37 @@ class VecSimTieredIndex : public VecSimIndexInterface {
...
+                                                     VecSimQueryParams *queryParams) override {
+        // Will be used only if a processing stage is needed
+        char processed_blob[this->backendIndex->getDataSize()];
+        const void *query_to_send = this->backendIndex->processBlob(queryBlob, processed_blob);
+
+        return this->rangeQuery(query_to_send, radius, queryParams);
+    }
+    virtual VecSimBatchIterator *
```
Comment (DvirDukhan) [inline src/VecSim/vec_sim_tiered_index.h:85]: new line

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**